functions/src/resource.ts

```diff
@@ -131,81 +149,107 @@ export class BaseResourceObject implements BaseResource {
+        }
+        const { user_id: userId, platform_id: platformId } = claims;
+        accessRules = [{type: "user", role: accessRuleRole, userId, platformId}] as UserAccessRule[];
+      }
+      else if (accessRuleType === "context") {
+        if (!accessRuleRole) {
+          reject("accessRuleRole fields missing!");
+          return;
+        }
+        if (!claims) {
+          reject("JWT claims missing!");
+          return;
+        }
+        const { platform_id: platformId, context_id: contextId } = claims;
+        if (!contextId) {
           reject("Missing context_id claim in JWT!");
           return;
         }
+        accessRules = [{type: "context", role: accessRuleRole, contextId, platformId}] as ContextAccessRule[];
+      }
+      else if (accessRuleType === "readWriteToken") {
+        // Generating new readWriteTokens is not recommended, they've been implemented to support documents imported
+        // from the Document Store. ReadWriteTokenPrefix is necessary so auth methods can differentiate it from
+        // regular JWT token. "token-service-generated:" is added just in case we want to easily find tokens/resources
+        // generated by token-service itself or by Document Store migration.
+        const readWriteToken = ReadWriteTokenPrefix + "token-service-generated:" + crypto.randomBytes(128).toString('hex');
+        accessRules = [{type: "readWriteToken", readWriteToken}] as ReadWriteTokenAccessRule[];
+      }
 
-        const accessRules = accessRuleType === "context"
-          ? [{type: "context", role: accessRuleRole, contextId, platformId}] as ContextAccessRule[]
-          : [{type: "user", role: accessRuleRole, userId, platformId}] as UserAccessRule[];
-
-        return BaseResourceObject.GetResourceSettings(db, env, type, tool)
-          .then((settings) => {
-            let newResource: FireStoreResource;
-            switch (type) {
-              case "s3Folder":
-                const {bucket, folder, region} = settings;
-                newResource = {
-                  type: "s3Folder",
-                  tool,
-                  name,
-                  description,
-                  accessRules,
-                  bucket,
-                  folder,
-                  region
-                }
-                break;
-
-              case "iotOrganization":
-                if (!(tool === "dataFlow")) {
-                  reject(`Unknown iotOrganization tool: ${tool}`);
-                  return;
-                }
-                newResource = {
-                  type: "iotOrganization",
-                  tool,
-                  name,
-                  description,
-                  accessRules
-                }
-                break;
-
-              default:
-                reject(`Unknown resource type: ${type}`);
+      return BaseResourceObject.GetResourceSettings(db, env, type, tool)
+        .then((settings) => {
+          let newResource: FireStoreResource;
```

Comment:
I think we should allow tools to specify if they allow readWriteToken access rules. Otherwise someone could create a readWriteToken access to the glossary, vortex data, or an IoT System.

Perhaps this is a reason to add a new resource type of s3FolderShared (or something like that). Then this could be done by limiting the tool to a resource type. Which could then be used to improve the conditional on line 216: `if (!(tool === "dataFlow")) {`

Reply:
There's a new "allowedAccessRuleTypes" array in settings now. 

Line 216 has been removed as settings define both tool and resource type as you noticed. So the code will fail higher up.